File: native/ios/Auspice/Screens/Palm.swift
import SwiftUI

struct PalmScreen: View {
    @Environment(\.scenePhase) private var scenePhase
    @State private var camera = LandmarkSession(kind: .hand)
    @SavedPractice("palm.lines") private var lines = LineTraits()
    @SavedPractice("palm.features") private var features: PalmFeatures? = nil
    @State private var error: String?
    @SavedPractice("palm.captured") private var captured = false

    private let fingerNames = [
        "jupiter": "Index finger", "saturn": "Middle finger",
        "apollo": "Ring finger", "mercury": "Little finger"
    ]

    var body: some View {
        ScreenScaffold(
            eyebrow: t("practice.palm"),
            title: t("palm.title"),
            tagline: t("palm.tagline")
        ) {
            Panel {
                ZStack {
                    CameraView(session: camera.session)
                    LandmarkOverlay(points: camera.overlay, joined: true)
                    if !camera.detecting {
                        Text(t("palm.hint"))
                            .font(Typeface.serif(17))
                            .foregroundStyle(Palette.ink)
                            .padding(.horizontal, 14)
                            .padding(.vertical, 8)
                            .background(Capsule().fill(.black.opacity(0.45)))
                    }
                }
                .frame(height: 320)
                .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
                .overlay(alignment: .topTrailing) { CameraFlipButton(session: camera) }
                .overlay(
                    RoundedRectangle(cornerRadius: 16, style: .continuous)
                        .strokeBorder(camera.detecting ? Palette.goldLine : Palette.line, lineWidth: 1)
                )
                if let message = camera.message {
                    Text(message).font(Typeface.sans(13)).foregroundStyle(Palette.inkMute)
                }
                Button(action: read) {
                    Label(captured ? t("palm.readAgain") : t("palm.read"), systemImage: "hand.raised")
                }
                .buttonStyle(PrimaryButtonStyle())
                .disabled(!camera.detecting)
                .opacity(camera.detecting ? 1 : 0.5)
            }

            Panel(title: t("palm.lines")) {
                Text(t("palm.linesNote"))
                    .font(Typeface.serif(16))
                    .foregroundStyle(Palette.inkSoft)
                    .fixedSize(horizontal: false, vertical: true)
                lineChoice(l("Heart line ends"), ["index": "Under the index", "middle": "Under the middle", "between": "Between them"], $lines.heart)
                lineChoice(l("Head line"), ["straight": "Straight", "curved": "Curved"], $lines.head)
                lineChoice(l("Life line"), ["wide": "Sweeps wide", "close": "Hugs the thumb"], $lines.life)
                lineChoice(l("Fate line"), ["present": "Present", "absent": "Absent", "unsure": "Not sure"], $lines.fate)
            }

            if let error {
                Panel(title: t("common.notComputed")) {
                    Text(error).font(Typeface.serif(16)).foregroundStyle(Palette.inkSoft)
                }
            }

            if let features {
                Panel(title: t("palm.hand")) {
                    measure(l("Shape"), shapeWord(features.shape))
                    measure(l("Palm width to length"), String(format: "%.2f", features.palmRatio))
                    measure(l("Fingers to palm"), String(format: "%.2f", features.fingerRatio))
                    measure(l("Index to ring"), String(format: "%.2f", features.indexToRing))
                    measure(l("Thumb angle"), String(format: "%.0f°", features.thumbAngle))
                    measure(l("Openness"), String(format: "%.2f", features.openness))
                }

                Panel(title: t("palm.fingers")) {
                    ForEach(features.fingers) { finger in
                        HStack(spacing: 10) {
                            Text(l(fingerNames[finger.finger] ?? finger.finger))
                                .font(Typeface.serif(16))
                                .foregroundStyle(Palette.ink)
                            Spacer(minLength: 0)
                            Text(l(finger.length))
                                .font(Typeface.sans(13, weight: .semibold))
                                .foregroundStyle(Palette.gold)
                            Text(String(format: "%.2f", finger.ratioToSaturn))
                                .font(Typeface.sans(13))
                                .foregroundStyle(Palette.inkMute)
                        }
                        .padding(.vertical, 4)
                    }
                }

                Panel(title: t("palm.mounts")) {
                    ForEach(features.palaces) { palace in
                        HStack(spacing: 10) {
                            Text(l(palace.palace))
                                .font(Typeface.serif(17))
                                .foregroundStyle(palace.state == "full" ? Palette.gold : Palette.ink)
                                .frame(width: 72, alignment: .leading)
                            GeometryReader { geometry in
                                ZStack(alignment: .leading) {
                                    Capsule().fill(Color.white.opacity(0.06))
                                    Capsule()
                                        .fill(palace.state == "full" ? Palette.gold : Palette.inkMute)
                                        .frame(width: geometry.size.width * min(1, max(0.02, (palace.prominence + 0.1) / 0.2)))
                                }
                            }
                            .frame(height: 7)
                            Text(l(palace.state))
                                .font(Typeface.sans(12))
                                .foregroundStyle(Palette.inkMute)
                                .frame(width: 42, alignment: .trailing)
                        }
                        .padding(.vertical, 3)
                    }
                    if !features.strongPalaces.isEmpty {
                        Text(lf("Standing out: {0}", features.strongPalaces.map(l).joined(separator: " · ")))
                            .font(Typeface.serif(16))
                            .foregroundStyle(Palette.gold)
                            .padding(.top, 4)
                    }
                }

                Panel(title: t("common.method")) {
                    Text(l("Proportions follow classical palmistry: the palm is square when its width reaches 0.86 of its length, the fingers long at 0.78 of the palm, and each finger is measured against the middle one. The mounts come from how far each stands out of the palm plane, which the landmarker reports as depth."))
                        .font(Typeface.serif(16))
                        .foregroundStyle(Palette.inkSoft)
                        .fixedSize(horizontal: false, vertical: true)
                }
                ExplainReading(result: features)
            }
        }
        .onAppear { Router.shared.palm = features; camera.start() }
        .onDisappear { camera.stop() }
        .onChange(of: scenePhase) { _, phase in
            if phase == .active { camera.start() } else { camera.stop() }
        }
    }

    private func lineChoice(_ label: String, _ options: [String: String], _ binding: Binding<String>) -> some View {
        VStack(alignment: .leading, spacing: 7) {
            FieldLabel(label)
            FlowRow(spacing: 8) {
                ForEach(options.sorted(by: { $0.key < $1.key }), id: \.key) { key, title in
                    Chip(label: l(title), active: binding.wrappedValue == key) {
                        binding.wrappedValue = key
                        if captured { read() }
                    }
                }
            }
        }
    }

    private func measure(_ label: String, _ value: String) -> some View {
        HStack {
            Text(l(label)).font(Typeface.serif(16)).foregroundStyle(Palette.inkSoft)
            Spacer()
            Text(l(value)).font(Typeface.sans(15, weight: .semibold)).foregroundStyle(Palette.ink)
        }
        .padding(.vertical, 3)
    }

    private func shapeWord(_ shape: String) -> String {
        l(shape.capitalized)
    }

    private func read() {
        let points = camera.landmarks
        guard points.count >= 21 else { error = l("No hand is in view."); return }
        do {
            let measured = try Engines.shared.evaluate(
                "palm.features",
                ["landmarks": points, "lines": lines.dictionary],
                as: PalmFeatures.self
            )
            features = measured
            Router.shared.palm = measured
            error = nil
            captured = true
        } catch {
            self.error = l("This reading could not be computed. Please try again.")
        }
    }
}

File: native/android/app/src/main/kotlin/art/lazying/auspice/screens/Palm.kt
package art.lazying.auspice.screens

import android.Manifest
import android.content.pm.PackageManager
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalLifecycleOwner
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.navigation.NavController
import art.lazying.auspice.*
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

private val FINGER_NAMES = mapOf(
    "jupiter" to "Index finger", "saturn" to "Middle finger",
    "apollo" to "Ring finger", "mercury" to "Little finger"
)

@Composable
fun PalmScreen(navController: NavController) {
    val context = LocalContext.current
    val owner = LocalLifecycleOwner.current
    val session = remember { LandmarkSession(hands = true) }
    var granted by remember {
        mutableStateOf(
            ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED
        )
    }
    val ask = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { granted = it }
    LaunchedEffect(Unit) { if (!granted) ask.launch(Manifest.permission.CAMERA) }
    DisposableEffect(Unit) { onDispose { session.stop() } }

    var lines by rememberPracticeState("palm.lines", LineTraits())
    var features by rememberPracticeState<PalmFeatures?>("palm.capture", null)
    var error by remember { mutableStateOf<String?>(null) }
    var reads by remember { mutableIntStateOf(0) }
    LaunchedEffect(features) { Router.palm = features }

    LaunchedEffect(reads) {
        if (reads == 0) return@LaunchedEffect
        val frames = session.captureFrames()
        if (frames.isEmpty()) { error = t("vision.hold"); return@LaunchedEffect }
        runCatching {
            Engines.evaluateAs<PalmFeatures>("palm.capture", buildJsonObject {
                put("frames", frames)
                put("lines", buildJsonObject {
                    put("heart", JsonPrimitive(lines.heart))
                    put("head", JsonPrimitive(lines.head))
                    put("life", JsonPrimitive(lines.life))
                    put("fate", JsonPrimitive(lines.fate))
                })
            })
        }.onSuccess { features = it; Router.palm = it; error = null }.onFailure { error = visionError(it) }
    }

    ScreenScaffold(
        navController,
        eyebrow = t("practice.palm"),
        title = t("palm.title"),
        tagline = t("palm.tagline")
    ) {
        Panel {
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(320.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .border(
                        1.dp,
                        if (session.detecting) Palette.goldLine else Palette.line,
                        RoundedCornerShape(16.dp)
                    ),
                contentAlignment = Alignment.Center
            ) {
                if (granted) {
                    CameraPreview(session, owner, Modifier.fillMaxSize())
                    LandmarkOverlay(session.overlay, joined = true, modifier = Modifier.fillMaxSize(), imageAspect = session.imageAspect)
                    if (!session.detecting) {
                        Text(
                            t("palm.hint"),
                            style = Type.serif(17), color = Palette.ink,
                            modifier = Modifier
                                .clip(CircleShape)
                                .background(Color(0x73000000))
                                .padding(horizontal = 14.dp, vertical = 8.dp)
                        )
                    }
                    CameraFlipButton(session, owner, Modifier.align(Alignment.TopEnd))
                } else {
                    Text(
                        t("camera.needed"),
                        style = Type.serif(17), color = Palette.inkSoft, textAlign = TextAlign.Center
                    )
                }
            }
            session.message?.let { Text(it, style = Type.sans(13), color = Palette.inkMute) }
            Text(t(if(session.ready) "vision.ready" else "vision.hold"), style = Type.sans(13), color = Palette.inkMute)
            PrimaryButton(
                if (features == null) t("palm.read") else t("palm.readAgain"),
                enabled = session.ready
            ) { reads++ }
            Text(t("face.privacy"), style = Type.sans(12), color = Palette.inkMute)
        }

        Panel(title = t("palm.lines")) {
            Text(
                t("palm.linesNote"),
                style = Type.serif(16), color = Palette.inkSoft
            )
            LineChoice(l("Heart line ends"), listOf("unsure" to "Not sure", "index" to "Under the index", "middle" to "Under the middle", "between" to "Between them"), lines.heart) {
                lines = lines.copy(heart = it); features = features?.copy(lines = lines); Router.palm = features
            }
            LineChoice(l("Head line"), listOf("unsure" to "Not sure", "straight" to "Straight", "curved" to "Curved"), lines.head) {
                lines = lines.copy(head = it); features = features?.copy(lines = lines); Router.palm = features
            }
            LineChoice(l("Life line"), listOf("unsure" to "Not sure", "wide" to "Sweeps wide", "close" to "Hugs the thumb"), lines.life) {
                lines = lines.copy(life = it); features = features?.copy(lines = lines); Router.palm = features
            }
            LineChoice(l("Fate line"), listOf("present" to "Present", "absent" to "Absent", "unsure" to "Not sure"), lines.fate) {
                lines = lines.copy(fate = it); features = features?.copy(lines = lines); Router.palm = features
            }
        }

        error?.let { Panel(title = t("common.notComputed")) { Text(it, style = Type.serif(16), color = Palette.inkSoft) } }

        features?.let { f ->
            Panel(title = t("vision.measurement")) {
                Text(t(if(f.measurement == null) "vision.legacy" else "vision.measured"), style=Type.sans(14), color=Palette.inkSoft)
                f.measurement?.takeIf { it.typeCandidates.size>1 }?.let { m ->
                    Text(m.typeCandidates.joinToString(" / ") { l(it.replaceFirstChar { c->c.uppercase() }) },style=Type.serif(17),color=Palette.gold)
                    Text(t("vision.boundary"),style=Type.sans(13),color=Palette.inkMute)
                }
            }
            Panel(title = t("palm.hand")) {
                Measure(l("Shape"), if(f.shape=="mixed") t("vision.mixed") else l(f.shape.replaceFirstChar { it.uppercase() }))
                Measure(l("Palm width to length"), String.format("%.2f", f.palmRatio))
                Measure(l("Fingers to palm"), String.format("%.2f", f.fingerRatio))
                Measure(l("Index to ring"), String.format("%.2f", f.indexToRing))
                Measure(l("Thumb angle"), String.format("%.0f°", f.thumbAngle))
                Measure(l("Openness"), String.format("%.2f", f.openness))
            }
            Panel(title = t("palm.fingers")) {
                f.fingers.forEach { finger ->
                    Row(Modifier.fillMaxWidth().padding(vertical = 4.dp), verticalAlignment = Alignment.CenterVertically) {
                        Text(l(FINGER_NAMES[finger.finger] ?: finger.finger), style = Type.serif(16), color = Palette.ink)
                        Spacer(Modifier.weight(1f))
                        Text(visionState(finger.length), style = Type.sans(13, FontWeight.SemiBold), color = Palette.gold)
                        Spacer(Modifier.width(8.dp))
                        Text(String.format("%.2f", finger.ratioToSaturn), style = Type.sans(13), color = Palette.inkMute)
                    }
                }
            }
            Panel(title = t("common.method")) {
                Text(t("vision.palmMethod"),
                    style = Type.serif(16), color = Palette.inkSoft
                )
            }
            ExplainReading(Json.encodeToString(f))
        }
    }
}

@Composable
private fun LineChoice(label: String, options: List<Pair<String, String>>, selected: String, onPick: (String) -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(7.dp)) {
        FieldLabel(label)
        FlowRowOf {
            options.forEach { (key, title) ->
                Chip(l(title), null, selected == key) { onPick(key) }
            }
        }
    }
}

@Composable
fun Measure(label: String, value: String) {
    Row(Modifier.fillMaxWidth().padding(vertical = 3.dp), verticalAlignment = Alignment.CenterVertically) {
        Text(l(label), style = Type.serif(16), color = Palette.inkSoft)
        Spacer(Modifier.weight(1f))
        Text(l(value), style = Type.sans(15, FontWeight.SemiBold), color = Palette.ink)
    }
}
